Fast modulus computation in MHash

CuriousGeorgiy opened the pull request:
# salad: refactor MHash

MHash had a redundant `n_buckets` field, which should always be calculated
from its `prime` field. But there was an inconsistency between them
introduced by `start_resize`: the `prime` field of both the real and
shadow table was updated, while only the shadow table's `n_buckets` field
was updated — apparently, the `prime` field of the real table should also
remain the same.

Fix the described inconsistency and remove the `n_buckets` field from
MHash.

NO_CHANGELOG=refactoring
NO_DOC=refactoring
NO_TEST=refactoring

# salad: fast modulus computation in MHash

Implement fast modulus computation in MHash based on algorithm described
in https://arxiv.org/abs/1902.01961.

Add a micro-benchmark for MHash: it shows a 10% performance improvement
with fast modulus computation enabled.

NO_CHANGELOG=optimization
NO_DOC=optimization
NO_TEST=optimization

CuriousGeorgiy committed:
salad: refactor MHash

MHash had a redundant `n_buckets` field, which should always be calculated
from its `prime` field. But there was an inconsistency between them
introduced by `start_resize`: the `prime` field of both the real and
shadow table was updated, while only the shadow table's `n_buckets` field
was updated — apparently, the `prime` field of the real table should also
remain the same.

Fix the described inconsistency and remove the `n_buckets` field from
MHash.

NO_CHANGELOG=refactoring
NO_DOC=refactoring
NO_TEST=refactoring

kyukhin commented:
Please, fix linter.

CuriousGeorgiy commented:
@kyukhin there is a wrong check in the linter, which needs to be disabled or worked-around, reported it to @locker.

CuriousGeorgiy committed:
perf: add benchmark for MHash

Add benchmark for MHash: this first version is quite basic and only
benchmarks random `find`s — needed for measuring performance gained from
fast modulus computation in a future patch.

NO_CHANGELOG=benchmark
NO_DOC=benchmark
NO_TEST=benchmark

Korablev77 commented:
Benchmark sometimes fails with
`Can't allocate 12884901928 bytes at /tarantool/src/lib/salad/mhash.h:530`

Korablev77 commented on src/lib/salad/mhash.h:
```diff
@@ -75,6 +79,26 @@
 
 #define mh_unlikely(x)  __builtin_expect((x),0)
 
+#define mh_mmi(p) (UINT64_C(0xFFFFFFFFFFFFFFFF) / (p) + 1)
```
What mmi stands for?

CuriousGeorgiy replied:
https://en.wikipedia.org/wiki/Modular_multiplicative_inverse

Korablev77 commented on src/lib/salad/mhash.h:
```diff
@@ -75,6 +79,26 @@
 
+#define mh_mmi(p) (UINT64_C(0xFFFFFFFFFFFFFFFF) / (p) + 1)
+
+#if MH_FAST_MOD
```
Personally I'd add fast mod to some util files and cover them with unit tests. Mb it turns out to be useful not only for mhash.

CuriousGeorgiy replied:
I believe **mhash** is the only appropriate use case in our codebase[^1]:
> Use cases: libdivide helps optimize the case of dividing many integer numerators by one integer denominator, where the denominator may only be known at runtime. There are many potential use cases: hash tables, scaling algorithms, a quick-'n-dirty speed boost for an interpreter, etc.

Let's consult @alyapunov.

[^1]: http://libdivide.com/

Korablev77 commented on src/lib/salad/mhash.h:
```diff
@@ -75,6 +79,26 @@
+
+#if MH_FAST_MOD
+#define mh_mul_u64_u32(u64, u32) \
+	({ (uint32_t)(((__uint128_t)(u64) * (u32)) >> 64); })
```
__uint128_t is extension and may be not supported by compiler. I guess we need extra checks here.

CuriousGeorgiy replied:
This compiles to **GCC** version as old as [4.1.2](https://godbolt.org/z/oaoo9jqTs), while the oldest compiler we support is **GCC** 4.8 on CentOS, so I believe this is nitpicking.

CuriousGeorgiy replied:
Fixed: added a conditional compilation check for this compiler feature.

Korablev77 reviewed (commented):
>Micro-benchmark for MHash shows up to 200% performance improvement on
random `find`s with fast modulus computation enabled.

Do you mean benchmark introduced in the first commit?

>Implement fast modulus computation in MHash based on algorithm based on
https://arxiv.org/abs/1902.01961.

Please extract main formulas to commit message so that others don't have to download article and search through 20 pages :)

CuriousGeorgiy commented on perf/mhash.cc:
```diff
@@ -0,0 +1,81 @@
+		::mh_u64_delete(h);
+	}
+
+protected:
```
Please take a look at `BENCHMARK_DEFINE_F` macro.

CuriousGeorgiy commented on perf/mhash.cc:
```diff
@@ -0,0 +1,81 @@
+	generate_rand_keys()
+	{
+		std::generate(rand_keys.begin(), rand_keys.end(), []() {
+			return filler_keys_count +
```
Yes. What seems weird to you? This is an abstraction over the generation of random keys. For now it's `rand` from **libc**, in the future it could be the [C++11 random library](https://en.cppreference.com/w/cpp/numeric/random).

CuriousGeorgiy replied:
Turned out the confusion was caused by the fact I actually test `find` over _absent_ keys. Supplied a comment and changed the naming correspondigly.

CuriousGeorgiy commented:
@Korablev77 
>Please extract main formulas to commit message so that others don't have to download article and search through 20 pages :)

I don't quite understand you. It's kind of hard to express math formulas in plain text — besides, they are the subject of the patch (i.e., can be found in code).

CuriousGeorgiy commented on perf/mhash.cc:
```diff
@@ -0,0 +1,81 @@
+	state.SetItemsProcessed(iterations_count);
+}
+
+BENCHMARK_REGISTER_F(MHashU64Fixture, RandomFind)->RangeMultiplier(2)->Range(1, 1 << 30);
```
Right, thanks! Fixed: exposed `max_keys_count` in fixture class and resused it here.

CuriousGeorgiy commented:
@Korablev77 
> Do you mean benchmark introduced in the first commit?

Thanks! Mentioned the corresponding commit in the message for clarity.

CuriousGeorgiy commented on perf/mhash.cc:
```diff
@@ -0,0 +1,81 @@
+#define MH_SOURCE
+#include <salad/mhash.h>
+
+class MHashU64Fixture : public benchmark::Fixture {
```
Fixed: changed the mapping to **uint64_t** -> **void \***.

CuriousGeorgiy committed:
salad: fast modulus computation in MHash

Implement fast modulus computation in MHash based on algorithm based on
https://arxiv.org/abs/1902.01961.

TLDR
`mod = dividend % divisor` can be computed by the following formulas:

`mmi = (UINT64_C(0xFFFFFFFFFFFFFFFF) / divisor + 1);`
`mod = (uint32_t)(((__uint128_t)(dividend * mmi) * divisor) >> 64);`

where mmi is the modular multiplicative inverse.

Micro-benchmark for MHash (b28abd00) shows up to 200% performance
improvement on random `find`s with fast modulus computation enabled.

NO_CHANGELOG=optimization
NO_DOC=optimization
NO_TEST=optimization
